### ISSUE TITLE
iMac didn't wake from sleep to print swatch.

### ISSUE THREAD

**Issue by appfollower** [jt70471/InkjetPlumber#8]
Using newest version  -- 9.4, 50 hour cycle.

**Comment by jt70471** [comment]
Hi Russ, when it did wake up, did it print the page? Trying to decided if it's just a wake-up issue or whether it failed to print altogether after you woke up your iMac.

Thanks

**Comment by appfollower** [comment]
Jeff:
I updated to 0.9.4 around 8:30 this morning, and put the computer back to sleep. I/P was scheduled to swatch print at 9:46. It didn't, so I assume it didn't wake the iMac. When I manually woke it a few minutes ago, it didn't signal the printer to print.
Here's what I see when I go into I/P:
<img width="787" alt="screen shot 2016-07-30 at 3 51 38 pm" src="https://cloud.githubusercontent.com/assets/20591358/17273564/8f515fea-566e-11e6-952d-e9f9b99b6b8f.png">
<img width="724" alt="screen shot 2016-07-30 at 3 52 28 pm" src="https://cloud.githubusercontent.com/assets/20591358/17273566/9655493c-566e-11e6-8d78-f2020543c2bf.png">
For some reason the Printers panel in Preferences is showing the last maintenance job as 12:27:33. Now, I don't want to complicate things, but when I came into my computer room a little while ago, there's a printed Canon nozzle check page sitting in my Pro100 tray. I definitely didn't command that. The nozzle check pages don't have a print time on them so I can't definitively connect it to the 12:27:33 print time in the Preferences panel, but it seems like more than a coincidence.
Russ

**Comment by appfollower** [comment]
I'm wondering now if having my Brother MFC490CW printer in the mix is affecting the program. Should I take it out?
Russ

**Comment by jt70471** [comment]
Hi Russ, the 0.9.4 release switched from printing the color swatches to the actual print driver's "self test page" functionality. I will likely offer both as options to choose in the preferences, but as of 0.9.4 the color swatches are not printed (nor an option to print).

I've been reading up on how to schedule the mac to wakup up at a future date/time and it's unfortunately pretty complicated. Not saying I can't get that done, but it's not something that is trivial to implement.

So, when you manually woke up your app, the maintenance jobs were overdue, which ran them shortly after you woke up your mac. The nozzle check you found on the output tray is indeed from one sent by Inkjet Plumber. Did your Brother also have a "self test page" of some sort in its output tray?

So, to properly implement the wake up functionality, it will take me some time to figure all of that out. Another option would be for IJP to prevent your computer from sleeping, but I don't want to go down that route, etc.

**Comment by appfollower** [comment]
No, nothing in the Brother print tray, Jeff.  I’m not too worried about that, since we just use it for random printing and the waste ink pad Canon problem isn’t an issue with the Brother. Its software doesn’t have all the test and clean options of the Canon. I was just doing it for the novelty.

What I’m confused about is that the ‘self test page’ that was in the Canon print tray when I came in seems to have been generated at a time when I was outside and the computer was in sleep mode. Sorry I didn’t monitor this more closely, but Im pretty sure I didn’t wake the computer just prior to the page being printed. From what you say, though, I gather there's simply  no possibility the page would have printed unless I woke it myself.

Russ

**Comment by jt70471** [comment]
I believe that's correct, although I do seem to recall you had a positive confirmation that the first version or so printed the color swatches when your computer had been asleep. Not sure exactly how that would be possible given what I'm now reading is required to wake the computer.

OK on the Brother, I've been testing with my Brother MFC-9460CDN and it does print a test page successfully. Perhaps the driver for yours doesn't provide that "print self test" page capability, which is another good argument for offering the choice of the "print self test" page or the original color swatches and that choice should be by printer I think...

**Comment by jt70471** [comment]
Russ, I just create 0.9.5 that gives you a preference setting to choose between the CUPS output (nozzle check) and the original custom color swatch page I was generating in earlier versions.

IJP should automatically inform you that an update is available, once you've had a chance to test choosing between CUPS or color swatch output, let me know. Thanks.

**Comment by appfollower** [comment]
Jeff:

The color swatch ran fine on the Brother, and the nozzle check page ran fine on the Canon. I’ll try the CUPS selection on the Brother later to see what it prints. I had a look, and there is a print test page for the Brother so maybe that’s what it will output.

A couple of things:

In preferences there’s an Auto Check for Updates box, which I think I’ve had checked in the past but didn’t get the auto updates. When I run the Check for Updates selection and there is an update there’s also a window that says Check for Updates Automatically. Do they both have to be selected to get the auto updates?

Also, in Preferences, when I set the print schedule for a specific printer and check OK, the window closes and I have to reopen it to change the schedule for other printers on the list. It’s a small thing, but it would be nifty if I could change the schedule for each printer on the list without the window closing and having to be reopened.

I really like the very clean appearance of the program.

Russ

**Comment by jt70471** [comment]
Thanks Russ. I am looking into options for allowing you to change more than one printer's preferences at a time, then clicking OK to save the settings for all of the printers, but as you said, right now, only the printer selected is saved and the dialog is dismissed when you click on OK.

Both my iMac and Macbook Pro auto-updated to 0.9.5 ok... The check for updates should happen every hour while the "Include development releases" preference is checked, otherwise it should check every 24 hours.

If you click on the "Check for Update" menu item, it will check immediately. I uploaded the 0.9.5 release at approximately 21:56 CDT and from my webserver log, I see a check for updates that occurred at 21:35 CDT which I believe was your client via the auto check. It looks like you may have manually checked for updates at 22:01 CDT, because that's when 0.9.5 was downloaded.

Of course I want to ensure that auto-update is working, but in this case from what I can tell, you would have gotten the auto-update at 22:35 CDT, etc.

I also have that "Automatically check and download" checkbox that shows up on the update notification. I'm working to figure out how to suppress or remove that, because my intention is that the "Automatically check for updates" in the preferences is intended to turn on/off that auto-update check.

Just one of the problems I haven't sorted yet, and normal for making pre-release development versions available. For a program like this, I think it's important to make early versions available, because there are tons of different printers out there and I have just 2 that I can test with.

I really appreciate all of your efforts in helping test IJP!

**Comment by appfollower** [comment]
OK, that makes sense.

I had the CUPS selected for my Brother. and it just now sent the command to the printer (the printer woke up) but nothing printed. I’ve reset it to color swatch to see if that works. The Canon page went through fine.

Russ

**Comment by jt70471** [comment]
Thanks for trying the CUPS option on your Brother. Seems not all printer drivers are going to implement the PrintSelfTestPage CUPS command, so glad I've added the option to choose between the two... :)

**Comment by appfollower** [comment]
Yes, just confirmed tonight CUPS selection doesn’t work on the Brother MFC490CW. Also confirmed the program does not wake the computer from sleep to send the print job. I was mistaken when I said it did. Sorry.
FYI, I also note that when the scheduled print is delayed due to the computer not being awake, when the computer is awakened and prints the page, that delay time amount is not added to the next scheduled print time. 

Russ

**Comment by jt70471** [comment]
Thanks Russ. The wakeup functionality will take some time to figure out, so you'll have to prevent your Mac from sleeping so that Inkjet Plumber can do its thing on schedule.

I'll check on the issue of not correctly determining the next schedule print if you have woken your mac past it's originally scheduled print. It should of course, reset the interval after it has printed the page, regardless of whether that occurred on schedule or later, etc.

**Comment by appfollower** [comment]
Jeff:
FYI, my computer definitely woke from sleep today and printed the Brother swatch and Pro100 nozzle check.
Russ

**Comment by jt70471** [comment]
That is pretty odd, I wonder if there may be something else that's waking your mac, because from everything I've read, the power management wake or wake and power on controls are only available to programs that run with elevated privileges, which Inkjet Plumber is not.

I still want to implement this functionality, but it will likely still be a few more weeks as I haven't had any time to work on this in the past week.

**Comment by appfollower** [comment]
I have it set for today around 12:30 to print. If it wakes up by itself I'll let you know. If it doesn't, I'll be completely befuddled and just put the whole subject aside.
Russ

**Comment by jt70471** [comment]
Everything I've read regarding programmatically waking a Mac from sleep indicates that you have to have a privileged helper program that runs with elevated privileged, which Inkjet Plumber doesn't do (yet).

In the past, when you have reported that the pages printed while sleeping, I suspect something else must have woken your mac, etc.

I am reviewing what I have to do to provide this functionality, but until I can get it implemented, you'll either have to prevent your mac from sleeping, i.e. you can blank the display, but don't sleep when on AC power, etc.
